### PR TITLE
[FLINK-3596] DataSet RelNode refactoring

### PR DESCRIPTION
**Pull request by vasia**
This PR refactors the current table API code so that

- the intermediate flink relnode layer and the dataset rules are removed
- code generation is moved from rules to DataSet nodes
- unused DataSete nodes are removed
- DataSetMap and DataSetReduce are merged into DataSetAggregate

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetAggregate.scala [hunk 1]
```diff
@@ -69,37 +72,55 @@ class DataSetGroupReduce(
...
     // get the output types
-    val fieldsNames = rowType.getFieldNames
     val fieldTypes: Array[TypeInformation[_]] = rowType.getFieldList.asScala
     .map(f => f.getType.getSqlTypeName)
     .map(n => TypeConverter.sqlTypeToTypeInfo(n))
     .toArray
 
     val rowTypeInfo = new RowTypeInfo(fieldTypes)
+
+    val mappedInput = inputDS.map(aggregateResult.mapFunc.apply(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetAggregate.scala]: We could refactor the `AggregateUtil` such that it returns the `MapFunction` directly and not a function to generate the `MapFunction`. I think would not need to expose the `intermediateDataType` of the `aggregateResult`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetAggregate.scala [hunk 2]
```diff
@@ -69,37 +72,55 @@ class DataSetGroupReduce(
...
+      config, inputDS.getType, intermediateType))
+
     val groupReduceFunction =
-      func.apply(config, inputDS.getType.asInstanceOf[RowTypeInfo], rowTypeInfo)
+      aggregateResult.reduceGroupFunc.apply(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetAggregate.scala]: `AggregateUtil` can directly return the `GroupReduceFunction`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCalc.scala [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  override def explainTerms(pw: RelWriter): RelWriter = {
+    super.explainTerms(pw).item("name", opName)
+  }
+
+  override def toString = opName
+
+  override def translateToPlan(config: TableConfig,
+      expectedType: Option[TypeInformation[Any]]): DataSet[Any] = {
+
+    expectedType match {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCalc.scala]: I think this match expression should be removed.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCalc.scala [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      case Some(typeInfo) if typeInfo.getTypeClass != classOf[Row] =>
+        throw new TableException("GroupReduce operations " +
+            "currently only support returning Rows.")
+      case _ => // ok
+    }
+
+    val inputDS = input.asInstanceOf[DataSetRel].translateToPlan(
+      config,
+      // tell the input operator that this operator currently only supports Rows as input
+      Some(TypeConverter.DEFAULT_ROW_TYPE))
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCalc.scala]: Remove this parameter.

**Comment by vasia** [comment]
Thanks for the review @fhueske. I've addressed your comments.

**Comment by fhueske** [comment]
Thanks for the update @vasia.
Merged to `tableOnCalcite`. PR can be closed.

**Comment by vasia** [comment]
Thanks for finalizing and merging :)

**Pull request closed without merging**